llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -1214,6 +1218,47 @@ InstructionCost VPWidenRecipe::computeCost(ElementCount VF,
   }
 }
 
+void VPWidenEVLRecipe::execute(VPTransformState &State) {
+  State.setDebugLocFrom(getDebugLoc());
+  assert(State.UF == 1 && "Expected only UF == 1 when vectorizing with "
+                          "explicit vector length.");
+  VPValue *Op0 = getOperand(0);
+
+  assert(State.get(Op0, 0)->getType()->isVectorTy() &&
+         "VPWidenEVLRecipe should not be used for scalars");
+
+  VPValue *EVL = getEVL();
+  Value *EVLArg = State.get(EVL, 0, /*NeedsScalar=*/true);
+  unsigned Opcode = getOpcode();
+  Instruction *I = getUnderlyingInstr();
+  IRBuilderBase &BuilderIR = State.Builder;
+  VectorBuilder Builder(BuilderIR);
+  Value *Mask = BuilderIR.CreateVectorSplat(State.VF, BuilderIR.getTrue());
+  Value *VPInst = nullptr;
+
+  //===------------------- Binary and Unary Ops ---------------------===//
+  if (Instruction::isBinaryOp(Opcode) || Instruction::isUnaryOp(Opcode)) {
+    // Just widen unops and binops.
+
+    SmallVector<Value *, 4> Ops;
+    for (unsigned I = 0, E = getNumOperands() - 1; I < E; ++I) {
+      VPValue *VPOp = getOperand(I);
+      Ops.push_back(State.get(VPOp, 0));
+    }
+
+    Builder.setMask(Mask).setEVL(EVLArg);
+    VPInst = Builder.createVectorInstruction(Opcode, Ops[0]->getType(), Ops,
+                                             "vp.op");
+
+    if (auto *VecOp = dyn_cast_or_null<Instruction>(VPInst))
+      VecOp->copyIRFlags(I);
```

Comment:
This should use the flags of the recipe, which may need to be dropped for correctness

```suggestion
      setFlags(VecOp);
```

See `@drop_vector_nuw_nsw` in `llvm/test/Transforms/LoopVectorize/X86/drop-poison-generating-flags.ll`, might be good to have a similar test for EVL widen recipes

Reply:
Current test won't be vectorized with EVL since it uses widen induction. https://github.com/llvm/llvm-project/pull/82021 is needed first to unblock widen EVL induction.
That said, can this test be added later ?

Reply:
Sure, test could also be added before widened EVL support.